core/trino-main/src/main/java/io/trino/server/security/oauth2/OAuth2Authenticator.java

```diff
@@ -77,4 +83,14 @@ protected AuthenticationException needAuthentication(ContainerRequestContext req
         URI tokenUri = request.getUriInfo().getBaseUri().resolve(getTokenUri(authId));
         return new AuthenticationException(message, format("Bearer x_redirect_server=\"%s\", x_token_server=\"%s\"", initiateUri, tokenUri));
     }
+
+    @Override
+    protected AuthenticationException needAuthentication(ContainerRequestContext request, String token, String message)
+    {
+        TokenPair tokenPair = tokenPairSerializer.deserialize(token);
+        Optional<UUID> refreshProcess = tokenRefresher.refreshToken(tokenPair);
+        return refreshProcess.map(refreshId -> request.getUriInfo().getBaseUri().resolve(getTokenUri(refreshId)))
+                .map(tokenUri -> new AuthenticationException(message, format("Bearer x_token_server=\"%s\"", tokenUri)))
```

Comment:
Ok so we can't do that I think. I don't believe that any of our client (jdbc/cli/python etc.) is taking the auth response headers and replacing it for the next request call. We do however allowed this `Bearer x_token_server` header to contain only x_token_server, so that client would only start polling for a new token to replace the old one, but would not try to open web-browser etc. This should be supported by every client that claims to support OAuth2/ExternalAuthentication protocol and it's the only way we can replace the existing token in these clients.  In a way, this is exactly what you are talking about, but executed in the norms of the already existing flow.

Reply:
My bad, I thought this worked differently and after digging into the code I realize that this is the right thing to do. It would be better if the client protocol was cleaner, but this is what we have right now.